Avatar performance

butonic opened the pull request:
revive of https://github.com/owncloud/core/pull/31105 this time with adjusted avatar migration, see https://github.com/owncloud/core/pull/32175 and https://github.com/owncloud/core/issues/28902 why it is needed for migrating user from 8.2 to 10, directly.

@tomneedham maybe you want to migrate the avatars completely differently?

cc @DeepDiver1975

butonic committed:
introduce IStorage based Node, Folder & File, make avatar manager work directly on storage to bypass cache

DeepDiver1975 commented:
conflicting + drone is red

PVince81 commented:
any update ?

PVince81 commented:
@butonic @tomneedham @DeepDiver1975 you guys please figure out how to continue this.

instructions for someone to take over ?

butonic commented:
I rebased. This PR introduces `OC\Files\Storage\Node`, `File` and `Folder` implementations that bypass the cache. In addition the `AvatarManager` passes the storage of the avatars folder to the `Avatar`, to 'unwrap' the `Folder` implementation it got. The `Avatar` than uses the cache bypassing `File` to do the actual filesystem interaction.

It is the followup for https://github.com/owncloud/core/pull/31105/files#r204521672 and also contains migration code. This is where @tomneedham may have to chime in because he had to deal with avatar migrations at length. Pun intended.

From my side if CI is green should be merged.

PVince81 commented:
since @DeepDiver1975 was familiar with this it would be good now to do one last review pass

tomneedham commented:
> @tomneedham maybe you want to migrate the avatars completely differently?

No this is ok. We had previous issues with <OC10 migration - unrealted to this.

tomneedham commented:
> @tomneedham maybe you want to migrate the avatars completely differently?

No this is ok. We had previous issues with <OC10 migration - unrealted to this. 

Only thing I would say is that we really only want to run the avatar migration once. Repair step is a bit heavy as it is executed on each upgrade step. Should maybe mark with a flag that this has been completed.

PVince81 commented:
ok. we can take over from here to just add a flag then.

thanks!

sharidas commented:
One idea is to introduce a flag in the config, after execution of method `moveAvatars()` in the file MoveAvatarOutsideHome.php, something like
```
$this->config->setSystemValue('avatarMigrationCompleted', 'true');
```
The reason I think is because avatar is not an app. So it would be nice to have a check if the config has `avatarMigrationCompleted` value `false` ( default val ) and then the version check (as in the current PR). I will create a PR based on this idea.

sharidas committed:
Add config flag to skip avatar repair step if already applied

Add config flag to skip avatar repair step if already applied.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

PVince81 commented:
@sharidas don't use `setSystemValue` as it would put it to config.php. Use the appconfig for core instead.

sharidas commented:
Updated the PR with appconfig, instead of using api which would put values into config.php.

PVince81 commented on lib/private/Repair/MoveAvatarOutsideHome.php:
```diff
@@ -146,9 +146,12 @@ private function moveAvatars(IOutput $out, IUser $user) {
+			\version_compare($ocVersionFromBeforeUpdate, '9.2.0.2', '<')) {
 			$function = function (IUser $user) use ($output) {
 				$this->moveAvatars($output, $user);
+				$this->config->setAppValue('core', 'avatarMigrationCompleted', 'true');
```
this would set the flag repeatedly for every user

sharidas replied:
:man_facepalming:  Ouch. I have updated the PR by moving this code after `$function` is executed.
Care is also taken to check if migration code is executed. Added doc in the code accordingly for the same.

sharidas commented:
Updated the PR. 
- Used another flag to check if the repair code is applied for at least one user.
- Set appconfig flag, if the repair code is executed. Not for each user, but after it is completed.

Let me know this looks ok. @PVince81

PVince81 commented on lib/private/Repair/MoveAvatarOutsideHome.php:
```diff
@@ -156,6 +165,11 @@ public function run(IOutput $output) {
 
+			//Set this if repair step is executed
+			if ($repairDoneForUsers === true) {
+				$this->config->setAppValue('core', 'avatarMigrationCompleted', 'true');
```
from what I see in oc_appconfig, the key name format is with underscores, not camel case. let's not start a new trend here

sharidas committed:
Lets not assume in the unit test

Lets not assume in the unit test. Assuming zeroth
index is file and first index is folder was failing
in the CI, intermittently.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

PVince81 commented on lib/private/Repair/MoveAvatarOutsideHome.php:
```diff
@@ -146,8 +146,17 @@ private function moveAvatars(IOutput $out, IUser $user) {
+			$repairDoneForUsers = false;
+			$function = function (IUser $user) use ($output, &$repairDoneForUsers) {
+				if ($repairDoneForUsers === false) {
+					$repairDoneForUsers = true;
```
why not just set it to true after `callForSeenUsers` instead of repeatedly ?

sharidas replied:
This iteration was caused because of my assumption with `callForSeenUsers`. I thought `callForSeenUsers` would be applied on users who logged in at least once. Unfortunately that's not the case. :man_facepalming: Thanks for the pointer.

PVince81 reviewed (approved):
:+1:

PVince81 commented:
@sharidas please backport

sharidas commented:
Backport PR: https://github.com/owncloud/core/pull/34592
